Add update-symlink argument to data download script

sjspielman opened the pull request:
Closes #530

This PR adds the `--update-symlink` argument to the `data-download.py` script, and it's working as expected on my end 👍 . I added a function `update_symlink()` which is called in two spots:
  - After downloading a data release (where the code originally was before I ported it into a function)
    - I also added a sentence here to print out how the symlink _would_ change if `--dryrun` is used 
  - Before any creds get used (to list releases), I include an `if args.update_symlink:` to update the symlink as desired. This is only done if the specified directory exists, and errors out otherwise with a message that you should download the data instead. I also use `sys.exit(0)` to exit out after changing the symlink, since we don't want to be doing any downloads with this flag.

Let me know if there are other behaviors we'd like to see for this flag. Also, I'm not convinced my help string for this argument is as clear as it could be, what do you think?

jashapiro commented on download-data.py:
```diff
@@ -244,12 +258,12 @@ def download_release_data(
+        if not dryrun:
+            update_symlink(download_dir, release)
+        else:
+            print(f"\nIf this were not a dry run, the 'current' symlink would be updated to point to '{release}'.")
```
I think we can be more compact here. 
```suggestion
            print(f"\nThe 'current' symlink would be updated to point to '{release}'.")
```

jashapiro commented on download-data.py:
```diff
@@ -346,7 +360,12 @@ def main() -> None:
+        "--update-symlink",
+        type=str,
+        default="",
+        help="The release version to update the 'current' symlink to direct to. Release directory must be present. Data will not be re-downloaded if it is already present."
```
This seemed to imply that data would be downloaded if not present.
```suggestion
        help="The release version to update the 'current' symlink to direct to. Release directory must be present. No data will be downloaded."
```

jashapiro reviewed (commented):
This looks mostly good, with a few wording suggestions and a suggestion to move the presence check to inside the function.

sjspielman commented:
I moved the check into the function, flipped the logic order, and also added some code to also print what would happen under the `--update_symlink <arg> --dryrun` flag combo.

jashapiro commented on download-data.py:
```diff
@@ -346,7 +370,12 @@ def main() -> None:
+        "--update-symlink",
+        type=str,
+        default="",
+        help="The release version to update the 'current' symlink to direct to. Release directory must be present. No data will be downloaded."
```
Do we want to support "latest" or a generic "release" value? I think we should! Let's not make people remember all of their release versions, and add a bit of logic to find the latest release.

jashapiro commented on download-data.py:
```diff
@@ -424,6 +453,15 @@ def main() -> None:
+
+    ### Update symlink if requested, without downloading data, if the target directory exists
+    if args.update_symlink:
+        if not args.dryrun:
```
Note that I also just learned about `.casefold()` which is a more robust alternative to `.lower()` for unicode: we could replace a number of calls to `.lower()` if we want to...

sjspielman commented:
I've updated the code to reflect the behavior described in https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/543#pullrequestreview-2146165879, and updated a few `.lower()` spots to `.casefold()`.
Any particular feedback on shorter ways to list directories, let me know. In listing local directories, I made an assumption (too strong?) that there is nothing in `data` besides release folders, but I could also list out the expected potential `regex`s for data folder names and find which are present?

Note that I test with these flags:
```
# works as expected
./download-data.py --update-symlink  
./download-data.py --update-symlink  --release 2024-05-01
./download-data.py --update-symlink  --release 2024-03-11 
./download-data.py --update-symlink  --release test
./download-data.py --update-symlink  --test-data

# fails as expected
./download-data.py --update-symlink --release 2024-10-11
```

jashapiro commented on download-data.py:
```diff
@@ -424,6 +461,37 @@ def main() -> None:
+        else:
+            if target == "current":
+                # find the most recent local release
+                dated_releases = [x for x in local_releases if re.match("\d{4}-\d{2}-\d{2}", x)]
```
move the `local_releases` here, because we don't need to look at all except when "current" is used. You can probably eliminate the variable completely by combining the two comprehensions:
```suggestion
                dated_releases = [x.name for x in args.data_dir.iterdir() if x.is_dir and re.match("\d{4}-\d{2}-\d{2}$", x.name)]
```

jashapiro commented on download-data.py:
```diff
@@ -424,6 +461,37 @@ def main() -> None:
+    if args.update_symlink:
+
+        # find all non-hidden directories in data_dir, representing the local releases
+        local_releases = [x.stem for x in args.data_dir.iterdir() if re.match("^[^.]", x.name) and x.is_dir()]
```
update: see below to remove/combine this.

jashapiro reviewed (commented):
I think we can simplify a but further! You are doing a number of duplicated checks, so I think we can rely on the checks in the function and eliminate a lot of extra nesting and more complex manipulation.

sjspielman commented:
:bowtie: https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/543/commits/716f981048c98e2dc28b0baa78c813c643090c93

jashapiro commented on download-data.py:
```diff
@@ -424,6 +461,26 @@ def main() -> None:
+
+        if target == "current":
+            # find the most recent local release to use as target
+            dated_releases = {x.name for x in args.data_dir.iterdir() if x.is_dir and re.match("\d{4}-\d{2}-\d{2}$", x.name)}
+            most_recent = sorted(dated_releases)[-1]
```
Since we are not using `in` or any other set operation, no need for a set here. (But since we only use it once, we can use a generator)
```suggestion
            dated_releases = (x.name for x in args.data_dir.iterdir() if x.is_dir and re.match("\d{4}-\d{2}-\d{2}$", x.name)
            most_recent = max(dated_releases)
```

jashapiro commented on download-data.py:
```diff
@@ -424,6 +461,26 @@ def main() -> None:
+
+        if target == "current":
+            # find the most recent local release to use as target
+            dated_releases = (x.name for x in args.data_dir.iterdir() if x.is_dir and re.match("\d{4}-\d{2}-\d{2}$", x.name)
```
I realized we probably want to check the date here too. (I didn't mean to send the last comment yet!)
```suggestion
            dated_releases = (
              x.name
              for x in args.data_dir.iterdir()
              if x.is_dir()
              and re.match("\d{4}-\d{2}-\d{2}$", x.name)
              and x.name <= datetime.date.today().isoformat()
            )
```

jashapiro reviewed (approved):
The last comment wasn't fully complete and had some errors, but I think what I have now should work...

With that change and local testing, this should be good to go.

sjspielman commented:
Noting that I ran the script locally with a variety of arguments, and it works as expected
